### ISSUE TITLE
False 'no-member' error for logbook - Pylint assumes incorrect instance type

### ISSUE THREAD

**Issue by ayalash** [PyCQA/pylint#3116]
When creating instance of `logbook.TestHnadler()`, pylint assumes that the type of this instance is `StackedObject` instead of `TestHandler`.
When the code tries to reach one of `TestHandler` members (declared in its `__init__` method), pylint reports "no-member" error.

### Steps to reproduce
Run pylint on:
```
$  cat /tmp/testme.py 
import logbook


with logbook.TestHandler() as handler:
    pass

print(handler.records)
```

### Current behavior
```
$ pylint /tmp/testme.py
************* Module testme
/tmp/testme.py:7:6: E1101: Instance of 'StackedObject' has no 'records' member (no-member)

--------------------------------------------------------------------
Your code has been rated at -2.50/10 (previous run: -2.50/10, +0.00)
```

### Expected behavior
No error

### pylint --version output
```
$ pylint --version
pylint 2.4.0
astroid 2.3.0
Python 3.6.9 (default, Jul  9 2019, 05:43:06) 
[GCC 9.1.0]
```

**Comment by ayalash** [comment]
This problem doesn't exist in older version of pylint/astroid:
```
$ pylint --version     
pylint 2.3.1
astroid 2.2.5
Python 3.6.9 (default, Jul  9 2019, 05:43:06) 
[GCC 9.1.0]
```

**Comment by s-kovacevic** [comment]
I'm getting false alarms for every dataclass attribute access. `astroid-2.3.0` introduced this error, works good with `2.2.0`. I'm assuming PyCQA/astroid#629 is causing the error.

**Comment by FrederikP** [comment]
I have the same issue as @ayalash and locking the astroid dependency to 2.2.5 and pylint to 2.3.1 fixed it for me. Only locking pylint to the older version didn't help, so the issue is probably in astroid, I guess.

**Comment by fvdnabee** [comment]
Same issue here with pylint 2.4.0 and astroid 2.3.0:
```
import numpy as np

def calc_mse(x1, x2):
    mse = np.square(x1 - x2).mean()
    return mse

X1 = np.arange(10)
X2 = X1 + 5

print(calc_mse(X1, X2))
```

Returns E1101: 
```
pylint_test.py:5:10: E1101: Function 'any' has no 'mean' member (no-member)
```

```pylint --version
pylint 2.4.0
astroid 2.3.0
Python 3.7.4 (default, Sep 11 2019, 01:11:53) 
[GCC 7.4.0]
```

Strangely enough I'm only able to reproduce this E1101 on an Ubuntu 18.04 machine (codeship). On Arch Linux with the same python, pylint and astroid versions (setup inside a virtualenv) this E1101 false positive doesn't pop up. I'm experiencing a similar issue where E1136 unsubscriptable-object errors only show up on the ubuntu codeship CI machine.  Perhaps a lib other than pylint and astroid is responsible? `pip freeze -l` is identical inside both virtualenvs though.

**Comment by PCManticore** [comment]
Folks, please stop commenting about false positives for different libraries, they need to be moved in separate issues, as each library might have a different root cause. For instance I can reproduce the OP's problem for `logrecord`, but not the one for `attrs` and `numpy`. Having them in separate issues means we can focus better on the root cause of each.

**Comment by PCManticore** [comment]
I believe @ayalash 's issue has been fixed with https://github.com/PyCQA/astroid/commit/130406463f3c978571a7aca6e33c985e17848f21. @ayalash Let me know if you can still reproduce this bug using `pylint` and `astroid` freshly installed from the Github repositories. We'll reopen if you can still reproduce it.
